File: python/liquid_crystal_print_task_manager.py
#!/usr/bin/python
# coding: UTF-8

import queue
import threading
import time
from lcd_core.liquid_crystal_mock_tk import LiquidCrystalMockTk
from liquid_crystal_virtual import LiquidCrystalVirtual
from liquid_crystal_print_task import LiquidCrystalPrintTask

class LiquidCrystalPrintTaskManager:
    def __init__(self, vlcd:LiquidCrystalVirtual, now, max_tasks=32):
        self.vlcd = vlcd
        self.max_tasks = max_tasks
        self.tasks = [LiquidCrystalPrintTask() for _ in range(self.max_tasks)]
        self.current_loop_time = now
        self._max_x = vlcd.max_x
        self._max_y = vlcd.max_y
    def _get_now_ms(self):
        return self.current_loop_time
    def _get_free_task(self):
        for t in self.tasks:
            if not t.is_active: return t
        return None
    def is_active_task_exist(self):
        for t in self.tasks:
            if t.is_active: return True
        return False
    def all_tasks_deactivate(self):
        for t in self.tasks:
            if t.is_active:
                t.is_active = False
    def all_tasks_erase(self):
        for t in self.tasks:
            if t.is_active:
                t.get_erase_data()

    def print(self, text, x, y, life_time, delay_time=0):
        t = self._get_free_task()
        if t == None: return
        t.text = text
        t.x = x
        t.y = y
        t.width = len(str(text))
        t.life_time = life_time
        t.delay_time = delay_time
        t.registered_time = self._get_now_ms()
        t.is_active = True
        self.vlcd.print(t.text, t.x, t.y)
    def blink(self, text, x, y, life_time, effect_speed=500, delay_time=0):
        t = self._get_free_task()
        if t == None: return
        t.text = text
        t.x = x
        t.y = y
        t.width = len(str(text))
        t.life_time = life_time
        t.delay_time = delay_time
        t.registered_time = self._get_now_ms()
        t.is_active = True
        t.effect_type = "BLINK"
        t.effect_speed = effect_speed
        self.vlcd.print(t.text, t.x, t.y)
    def slide(self, text, x, y, width, life_time, effect_speed=500, effect_param=1, delay_time=0):
        t = self._get_free_task()
        if t == None: return
        t.text = text
        t.x = x
        t.y = y
        t.width = width
        t.life_time = life_time
        t.delay_time = delay_time
        t.registered_time = self._get_now_ms()
        t.is_active = True
        t.effect_type = "SLIDE"
        t.effect_speed = effect_speed
        t.effect_param = effect_param
        self.vlcd.print(t.text, t.x, t.y)
    def knight(self, text, x, y, width, life_time, effect_speed=500, effect_param=1, delay_time=0):
        t = self._get_free_task()
        if t == None: return
        t.text = text
        t.x = x
        t.y = y
        t.width = width
        t.life_time = life_time
        t.delay_time = delay_time
        t.registered_time = self._get_now_ms()
        t.is_active = True
        t.effect_type = "KNIGHT"
        t.effect_speed = effect_speed
        t.effect_param = effect_param
        self.vlcd.print(t.text, t.x, t.y)

    def update(self, now):
        self.current_loop_time = now
        if not self.is_active_task_exist(): return

        for t in self.tasks:
            if not t.is_active: continue

            text, tx, ty = t.evaluate(now)
            self.vlcd.print(text, tx, ty)

def loop(data_queue, vlcd, pm):
    while True:
        now = int(time.time() * 1000)
        pm.update(now)

        vlcd.flush()
        data_queue.put("update")
        time.sleep(0.05)

if __name__ == "__main__":
    shared_queue = queue.Queue()
    size_x = 20
    size_y = 4
    lcd = LiquidCrystalMockTk(shared_queue, size_x, size_y, blank_ch="_")
    vlcd = LiquidCrystalVirtual(lcd, size_x=size_x, size_y=size_y)
    now = int(time.time() * 1000)
    pm = LiquidCrystalPrintTaskManager(vlcd, now)
    pm.print("Hello, World!", 0, 0, 3000, delay_time=10000)
    pm.blink("Blinking", 3, 1, 10000, effect_speed=500)
    pm.slide("Sliding", 0, 2, 20, 10000, effect_speed=100, effect_param=-1)
    pm.knight("Knight", 0, 3, 20, 10000, effect_speed=100, effect_param=-1)

    loop_thread = threading.Thread(
        target=loop, args=(shared_queue, vlcd, pm,), daemon=True
    )
    loop_thread.start()

    lcd.mainloop()

File: cpp/LiquidCrystalPrintTaskManager.cpp
#include "LiquidCrystalPrintTaskManager.h"
#include <string.h>
#include <stdio.h>

LiquidCrystalPrintTaskManager::LiquidCrystalPrintTaskManager(LiquidCrystalVirtual* virtual_lcd, uint32_t now, uint8_t max_t) {
    vlcd = virtual_lcd;
    max_tasks = max_t;
    tasks = new LiquidCrystalPrintTask[max_tasks];
    current_loop_time = now;
    _max_x = vlcd->max_x();
    _max_y = vlcd->max_y();
}

LiquidCrystalPrintTaskManager::~LiquidCrystalPrintTaskManager() {
    delete[] tasks;
}

LiquidCrystalPrintTask* LiquidCrystalPrintTaskManager::_get_free_task() {
    for (uint8_t i = 0; i < max_tasks; ++i) {
        if (!tasks[i].is_active) return &tasks[i];
    }
    return nullptr;
}

bool LiquidCrystalPrintTaskManager::is_active_task_exist() const {
    for (uint8_t i = 0; i < max_tasks; ++i) {
        if (tasks[i].is_active) return true;
    }
    return false;
}

void LiquidCrystalPrintTaskManager::all_tasks_deactivate() {
    for (uint8_t i = 0; i < max_tasks; ++i) {
        tasks[i].is_active = false;
    }
}

void LiquidCrystalPrintTaskManager::all_tasks_erase() {
    for (uint8_t i = 0; i < max_tasks; ++i) {
        if (tasks[i].is_active) {
            tasks[i].get_erase_data();
        }
    }
}

void LiquidCrystalPrintTaskManager::print(const char* text, int16_t x, int16_t y, uint32_t life_time, uint32_t delay_time) {
    LiquidCrystalPrintTask* t = _get_free_task();
    if (t == nullptr) return;
    
    snprintf(t->text, sizeof(t->text), "%s", text);
    t->x = x;
    t->y = y;
    t->width = strlen(text);
    t->life_time = life_time;
    t->delay_time = delay_time;
    t->registered_time = current_loop_time;
    t->effect_type = EffectType::NONE;
    t->is_active = true;
    
    vlcd->print(t->text, t->x, t->y);
}

void LiquidCrystalPrintTaskManager::blink(const char* text, int16_t x, int16_t y, uint32_t life_time, uint32_t effect_speed, uint32_t delay_time) {
    LiquidCrystalPrintTask* t = _get_free_task();
    if (t == nullptr) return;

    snprintf(t->text, sizeof(t->text), "%s", text);
    t->x = x;
    t->y = y;
    t->width = strlen(text);
    t->life_time = life_time;
    t->delay_time = delay_time;
    t->registered_time = current_loop_time;
    t->effect_type = EffectType::BLINK;
    t->effect_speed = effect_speed;
    t->is_active = true;

    vlcd->print(t->text, t->x, t->y);
}

void LiquidCrystalPrintTaskManager::slide(const char* text, int16_t x, int16_t y, uint8_t width, uint32_t life_time, uint32_t effect_speed, int8_t effect_param, uint32_t delay_time) {
    LiquidCrystalPrintTask* t = _get_free_task();
    if (t == nullptr) return;

    snprintf(t->text, sizeof(t->text), "%s", text);
    t->x = x;
    t->y = y;
    t->width = width;
    t->life_time = life_time;
    t->delay_time = delay_time;
    t->registered_time = current_loop_time;
    t->effect_type = EffectType::SLIDE;
    t->effect_speed = effect_speed;
    t->effect_param = effect_param;
    t->is_active = true;

    vlcd->print(t->text, t->x, t->y);
}

void LiquidCrystalPrintTaskManager::knight(const char* text, int16_t x, int16_t y, uint8_t width, uint32_t life_time, uint32_t effect_speed, int8_t effect_param, uint32_t delay_time) {
    LiquidCrystalPrintTask* t = _get_free_task();
    if (t == nullptr) return;

    snprintf(t->text, sizeof(t->text), "%s", text);
    t->x = x;
    t->y = y;
    t->width = width;
    t->life_time = life_time;
    t->delay_time = delay_time;
    t->registered_time = current_loop_time;
    t->effect_type = EffectType::KNIGHT;
    t->effect_speed = effect_speed;
    t->effect_param = effect_param;
    t->is_active = true;

    vlcd->print(t->text, t->x, t->y);
}

void LiquidCrystalPrintTaskManager::update(uint32_t now) {
    current_loop_time = now;
    if (!is_active_task_exist()) return;

    for (uint8_t i = 0; i < max_tasks; ++i) {
        if (!tasks[i].is_active) continue;

        EvaluateResult res = tasks[i].evaluate(now);
        vlcd->print(res.text, res.x, res.y);
    }
}












